sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuDeviceManager.scala

```diff
@@ -427,65 +427,75 @@ object GpuDeviceManager extends Logging {
-      logWarning("OFF HEAP MEMORY LIMITS IS ENABLED. " +
-          "THIS IS EXPERIMENTAL FOR NOW USE WITH CAUTION")
+    val (pinnedSize, nonPinnedLimit) = {
       val perTaskOverhead = conf.perTaskOverhead
       val totalOverhead = perTaskOverhead * GpuDeviceManager.numCores
+      val executorOverhead = conf.executorOverhead
       val confPinnedSize = conf.pinnedPoolSize
       val confLimit = conf.offHeapLimit
       // TODO the min limit size of overhead + 1 GiB is arbitrary and we should have some
       ///  better tests to see what an ideal value really should be.
       val minMemoryLimit = totalOverhead + (1024 * 1024 * 1024)
+      val pysparkOverhead = conf.pysparkOverhead
+      val heapSize = conf.heapSize
+      val deviceCount: Int = Cuda.getDeviceCount()
 
       val memoryLimit = if (confLimit.isDefined) {
+        if (executorOverhead.isEmpty) {
+          logWarning(s"${RapidsConf.EXECUTOR_OVERHEAD_SIZE} is not set")
+        }
         confLimit.get
-      } else if (confPinnedSize > 0) {
-        // TODO 1 GiB above the pinned size probably should change, we are using it to match the
-        //  old behavior for the pool, but that is not great and we want to have hard evidence
-        //  for a better value before just picking something else that is arbitrary
-        val ret = confPinnedSize + (1024 * 1024 * 1024)
-        logWarning(s"${RapidsConf.OFF_HEAP_LIMIT_SIZE} is not set using " +
-            s"${RapidsConf.PINNED_POOL_SIZE} + 1 GiB instead for memory limit " +
-            s"${ret / 1024 / 1024.0} MiB")
-        ret
-      } else {
-        // We don't have pinned or a conf limit set, so go with the min
-        logWarning(s"${RapidsConf.OFF_HEAP_LIMIT_SIZE} is not set and neither is " +
-            s"${RapidsConf.PINNED_POOL_SIZE}. Using the minimum memory size for the limit " +
-            s"which is ${RapidsConf.TASK_OVERHEAD_SIZE} * cores (${GpuDeviceManager.numCores}) " +
-            s"+ 1 GiB => ${minMemoryLimit / 1024 / 1024.0} MiB")
-        minMemoryLimit
-      }
-
-      val finalMemoryLimit = if (memoryLimit < minMemoryLimit) {
-        logWarning(s"The memory limit ${memoryLimit / 1024 / 1024.0} MiB " +
-            s"is smaller than the minimum limit size ${minMemoryLimit / 1024 / 1024.0} MiB " +
-            s"using the minimum instead")
-        minMemoryLimit
       } else {
-        memoryLimit
+        // in case we cannot query the host for available memory due to environmental
+        // constraints, we can fall back to minMemoryLimit via saying there's no available
+        lazy val availableHostMemory = MemoryChecker.getAvailableMemoryBytes.getOrElse(0L)
+        lazy val basedOnHostMemory = (.8 * (availableHostMemory - heapSize - pysparkOverhead) /
+          deviceCount).toLong
+        if (executorOverhead.isDefined) {
+          val basedOnConfiguredOverhead = (.9 * (executorOverhead.get - pysparkOverhead)).toLong
```

Comment:
Regarding Bobby's question of whether executor memory overhead is supposed to include python overhead, the spark config docs page says "it is up to the application to avoid exceeding the overhead memory space shared with other non-JVM processes" so I think the answer is yes.

https://spark.apache.org/docs/latest/configuration.html